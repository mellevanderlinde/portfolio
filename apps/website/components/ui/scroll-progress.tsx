'use client'

import type { SpringOptions } from 'motion/react'
import type { ReactNode, RefObject } from 'react'
import { cn } from '@/lib/utils'
import { motion, useScroll, useSpring } from 'motion/react'

interface ScrollProgressProps {
  className?: string
  springOptions?: SpringOptions
  containerRef?: RefObject<HTMLDivElement>
}

const defaultSpringOptions: SpringOptions = {
  stiffness: 200,
  damping: 50,
  restDelta: 0.001,
}

export function ScrollProgress({
  className,
  springOptions,
  containerRef,
}: ScrollProgressProps): ReactNode {
  const { scrollYProgress } = useScroll({
    container: containerRef,
    layoutEffect: Boolean(containerRef?.current),
  })

  const scaleX = useSpring(scrollYProgress, {
    ...defaultSpringOptions,
    ...(springOptions ?? {}),
  })

  return (
    <motion.div
      className={cn('inset-x-0 top-0 h-1 origin-left', className)}
      style={{
        scaleX,
      }}
    />
  )
}
