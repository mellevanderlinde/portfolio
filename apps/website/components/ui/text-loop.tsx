'use client'
import type {
  AnimatePresenceProps,
  Transition,
  Variants,
} from 'motion/react'
import type { ReactNode } from 'react'
import { cn } from '@/lib/utils'
import {
  AnimatePresence,
  motion,
} from 'motion/react'
import { useEffect, useState } from 'react'

interface TextLoopProps {
  children: ReactNode[]
  className?: string
  interval?: number
  transition?: Transition
  variants?: Variants
  onIndexChange?: (index: number) => void
  trigger?: boolean
  mode?: AnimatePresenceProps['mode']
}

const defaultTransition: Transition = { duration: 0.3 }

export function TextLoop({
  children,
  className,
  interval = 2,
  transition = defaultTransition,
  variants,
  onIndexChange,
  trigger = true,
  mode = 'popLayout',
}: TextLoopProps): ReactNode {
  const [currentIndex, setCurrentIndex] = useState(0)
  const items = [...children]

  useEffect(() => {
    if (!trigger)
      return

    const intervalMs = interval * 1000
    const timer = setInterval(() => {
      setCurrentIndex((current) => {
        const next = (current + 1) % items.length
        onIndexChange?.(next)
        return next
      })
    }, intervalMs)
    return () => clearInterval(timer)
  }, [items.length, interval, onIndexChange, trigger])

  const motionVariants: Variants = {
    initial: { y: 20, opacity: 0 },
    animate: { y: 0, opacity: 1 },
    exit: { y: -20, opacity: 0 },
  }

  return (
    <div className={cn('relative inline-block whitespace-nowrap', className)}>
      <AnimatePresence mode={mode} initial={false}>
        <motion.div
          key={currentIndex}
          initial="initial"
          animate="animate"
          exit="exit"
          transition={transition}
          variants={variants || motionVariants}
        >
          {items[currentIndex]}
        </motion.div>
      </AnimatePresence>
    </div>
  )
}
