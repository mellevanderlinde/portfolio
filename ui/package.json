{
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build"
  },
  "dependencies": {
    "@tailwindcss/postcss": "4.0.0-alpha.13",
    "@types/node": "^22.5.0",
    "@types/react": "^18.2.55",
    "@types/react-dom": "^18.2.19",
    "highlight.js": "^11.10.0",
    "next": "^14.2.7",
    "next-mdx-remote": "^5.0.0",
    "postcss": "^8.4.35",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "4.0.0-alpha.13",
    "typescript": "~5.6.2"
  }
}
