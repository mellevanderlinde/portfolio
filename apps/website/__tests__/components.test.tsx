import { AnimatedBackground } from '@/components/ui/animated-background'
import { Magnetic } from '@/components/ui/magnetic'
import { ScrollProgress } from '@/components/ui/scroll-progress'
import { Spotlight } from '@/components/ui/spotlight'
import { TextEffect } from '@/components/ui/text-effect'
import { TextLoop } from '@/components/ui/text-loop'
import { render } from '@testing-library/react'
import { expect, it } from 'vitest'

it('matches animated background', () => {
  const { container } = render(<AnimatedBackground children={<></>} />)
  expect(container).toMatchSnapshot()
})

it('matches magnetic', () => {
  const { container } = render(<Magnetic children={<></>} />)
  expect(container).toMatchSnapshot()
})

it ('matches scroll progress', () => {
  const { container } = render(<ScrollProgress />)
  expect(container).toMatchSnapshot()
})

it('matches spotlight', () => {
  const { container } = render(<Spotlight />)
  expect(container).toMatchSnapshot()
})

it('matches text effect', () => {
  const { container } = render(<TextEffect children="" />)
  expect(container).toMatchSnapshot()
})

it('matches text loop', () => {
  const { container } = render(<TextLoop children={[]} />)
  expect(container).toMatchSnapshot()
})
