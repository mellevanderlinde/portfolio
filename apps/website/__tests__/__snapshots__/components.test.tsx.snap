// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches album slider 1`] = `
<div>
  <main
    class="space-y-24"
    style="opacity: 0;"
  >
    <section
      class="mt-18"
      style="opacity: 0; filter: blur(8px); transform: translateY(20px);"
    >
      <div
        class="flex-1"
      >
        <p
          class="text-zinc-600 dark:text-zinc-400"
        >
          These are a few of my favorite albums. You can also find me on
           
          <u>
            <a
              href="https://soundcloud.com/melledj"
              rel="noopener noreferrer"
              target="_blank"
            >
              SoundCloud
            </a>
          </u>
          .
        </p>
      </div>
      <div
        class="overflow-hidden mt-8"
      >
        <div
          class="flex w-max"
          style="gap: 24px; flex-direction: row; transform: none;"
        >
          <img
            alt="Artist - Album"
            class="aspect-square w-[120px] rounded-[4px]"
            src="https://i.scdn.co/image/id"
          />
          <img
            alt="Artist - Album"
            class="aspect-square w-[120px] rounded-[4px]"
            src="https://i.scdn.co/image/id"
          />
        </div>
      </div>
    </section>
  </main>
</div>
`;

exports[`matches animated background 1`] = `
<div>
  <div
    class="z-10"
  />
</div>
`;

exports[`matches code 1`] = `
<div>
  <div>
    <pre
      class="shiki shiki-themes github-light github-dark-high-contrast"
      style="background-color:#fff;--shiki-dark-bg:#000000;color:#000000;--shiki-dark:#000000"
      tabindex="0"
    >
      <code>
        <span
          class="line"
        >
          <span
            style="color:#000000;--shiki-dark:#000000"
          >
            const
          </span>
          <span
            style="color:#000000;--shiki-dark:#000000"
          >
             a
          </span>
          <span
            style="color:#000000;--shiki-dark:#000000"
          >
             =
          </span>
          <span
            style="color:#000000;--shiki-dark:#000000"
          >
             1
          </span>
          <span
            style="color:#000000;--shiki-dark:#000000"
          >
            ;
          </span>
        </span>
      </code>
    </pre>
  </div>
</div>
`;

exports[`matches infinite slider 1`] = `
<div>
  <div
    class="overflow-hidden"
  >
    <div
      class="flex w-max"
      style="gap: 16px; flex-direction: row; transform: none;"
    />
  </div>
</div>
`;

exports[`matches scroll progress 1`] = `
<div>
  <div
    class="inset-x-0 top-0 h-1 origin-left"
    style="transform: scaleX(0);"
  />
</div>
`;

exports[`matches spotlight 1`] = `
<div
  style="position: relative; overflow: hidden;"
>
  <div
    class="pointer-events-none absolute rounded-full bg-[radial-gradient(circle_at_center,var(--tw-gradient-stops),transparent_80%)] blur-xl transition-opacity duration-200 from-zinc-50 via-zinc-100 to-zinc-200 opacity-0"
    style="width: 200px; height: 200px; left: -100px; top: -100px;"
  />
</div>
`;

exports[`matches text effect 1`] = `
<div>
  <p
    style="opacity: 0;"
  >
    <span
      class="sr-only"
    />
    <span
      aria-hidden="true"
      class="inline-block whitespace-pre"
      style="opacity: 0;"
    />
  </p>
</div>
`;

exports[`matches text loop 1`] = `
<div>
  <div
    class="relative inline-block whitespace-nowrap"
  >
    <div
      style="opacity: 1; transform: none;"
    />
  </div>
</div>
`;
